Adds the ability to use different payloads

mustiikhalil opened the pull request:
Good morning! 

The following PR will be moving the `Message Protocol` away from the code base and introduces the `Payload protocol` which will allow users, to decode messages according to the Type they want to send. it addresses the following issue: #708, @glbrntt let's see how we can make this better since I found the following issue:

- issues:
 - Adds an over head of the code gen since now you will have to add the
```swift
extension Helloworld_HelloReply: Payload {
    public init(serializedByteBuffer: inout ByteBuffer) throws {
        try self.init(serializedData: serializedByteBuffer.readData(length: serializedByteBuffer.readableBytes)!,
                  extensions: nil,
                  partial: false,
                  options: BinaryDecodingOptions())
    }
    
    public func serialize(into buffer: inout ByteBuffer) throws {
        let data = try serializedData()
        buffer.writeBytes(data)
    }
}
``` 
which is not needed all the time since the following object didn't require it to work
```swift
extension Routeguide_RouteNote: Payload {
    public init(serializedByteBuffer: inout ByteBuffer) throws {}
    public func serialize(into buffer: inout ByteBuffer) throws {}
}
```

mustiikhalil commented on Sources/GRPC/ReadWriteStates.swift:
```diff
@@ -57,23 +57,26 @@ enum WriteState {
-      var buffer = allocator.buffer(capacity: 0)
+      let data = buffer.readData(length: buffer.readableBytes)!
+      buffer.clear()
       writer.write(data, into: &buffer)
```
is there a prettier way of doing this?

glbrntt replied:
Yes, but unfortunately we need to rework the writer API slightly. I have a PR up at the moment which adds support for compression which complicates things a little more. For the time being I think making the writer generic over the payload and having it serialize the payload into the buffer is sufficient.

glbrntt commented on Sources/GRPC/Payload.swift:
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright YEARS, gRPC Authors All rights reserved.
```
This should be 2020 instead of YEARS (the script replaces acceptable years or ranges of years with "YEARS" since it's easier to check against a 'correct' header)

glbrntt commented on Sources/Examples/Echo/Model/echo.grpc.swift:
```diff
@@ -149,3 +149,12 @@ extension Echo_EchoProvider {
   }
 }
 
+extension Echo_EchoRequest: Payload {
```
This file is generated, but it looks like these were added manually.

The implementation for the conformance to `Payload` _shouldn't_ be in the generated code. Instead we can create another protocol (e.g. `GRPCProtobufPayload`) which conforms to `Payload` and `Message`. It should then provide the default implementation for `Payload` so that the generated messages can be extended to conform to that protocol.

You'll need to take a look at the code-gen (in `protoc-gen-grpc-swift`) to figure out how to do this.

mustiikhalil replied:
@glbrntt Yes, I just added them to see if they would pass the test cases and to have a draft of the PR. so we can talk about What we would be doing, and how.

we can have the following:
```swift
protocol GRPCProtobufPayload: Payload {}
extension GRPCProtobufPayload {
// comes the implementation of the functions
}
```
and similarly in FlatBuffers:
```swift
protocol GRPCFlatbufPayload: Payload {}
extension GRPCFlatbufPayload {
// comes the implementation of the functions
}
```
do you think we should generate the code? or implement it which the package itself?

glbrntt replied:
Ah okay, no problem. Just an FYI if you want to run the tests locally you can set the `ENABLE_TIMING_TESTS` environment variable to `false` to skip some of the longer running tests (it takes the test time down to a couple of seconds).

The default implementation for `GRPCProtobufPayload` (which should also conform to `SwiftProtobuf.Message`) should be within `GRPC` since we already depend on `SwiftProtobuf`. This gives us more flexibility to change the implementation (i.e. if/when new Protobuf APIs become available) without regenerating the code.

For flatbuffers I think we _have_ to generate the conformance to `GRPCPayload` each time since there's no dependency between flatbuffers and gRPC Swift here.

glbrntt commented on Sources/GRPC/Payload.swift:
```diff
@@ -0,0 +1,22 @@
+ * limitations under the License.
+ */
+import NIO
+import SwiftProtobuf
```
This isn't required here

glbrntt commented on Sources/GRPC/Payload.swift:
```diff
@@ -0,0 +1,22 @@
+public protocol Payload {
+    init(serializedByteBuffer: inout NIO.ByteBuffer) throws
+    func serialize(into buffer: inout NIO.ByteBuffer) throws
+}
```
We'll need some documentation here. It might also be worth actually naming this `GRPCPayload` instead, what do you think?

mustiikhalil replied:
Yeah that sounds good! `typealise GRPCPayload = NIO.ByteBuffer`

glbrntt replied:
Sorry, I meant rename the protocol `Payload` to `GRPCPayload`, not a type alias for `ByteBuffer`

mustiikhalil replied:
Makes sense

mustiikhalil commented on Sources/GRPC/GRPCPayload.swift:
```diff
@@ -14,9 +14,15 @@
+    }
+    
+    func serialize(into buffer: inout NIO.ByteBuffer) throws {}
 }
```
I updated the code as recommended however, now we are facing an issue with the current way it's implemented. the `GRPCProtoBufPayload` protocol functions aren't being called, and only the ones in the GRPCPayload are being called instead. Which means we are not serializing nor deserializing the buffer

mustiikhalil commented on Sources/GRPC/GRPCProtobufPayload.swift:
```diff
@@ -0,0 +1,32 @@
+        let data = try self.serializedData()
+        buffer.writeBytes(data)
+    }
+}
```
Never being called, since the code relies on the GRPCPayload instead of this protocol

glbrntt replied:
Added some comments to help fix this :)

glbrntt commented on Sources/GRPC/GRPCPayload.swift:
```diff
@@ -0,0 +1,28 @@
+    }
+    
+    func serialize(into buffer: inout NIO.ByteBuffer) throws {}
+}
```
This is partly where the problem is coming from, this should be:

```swift
public protocol GRPCPayload {
  init(serializedByteBuffer: inout NIO.ByteBuffer) throws
  func serialize(into buffer: inout NIO.ByteBuffer) throws
}
```

There _should not_ be a default implementation on the `GRPCPayload`.

glbrntt commented on Sources/GRPC/GRPCProtobufPayload.swift:
```diff
@@ -0,0 +1,32 @@
+import NIO
+import SwiftProtobuf
+
+public protocol GRPCProtobufPayload: GRPCPayload {}
```
`GRPCProtobufPayload` by definition must be a `SwiftProtobuf.Message`, I don't think there's any need to make the extension conditional on that, it might as well be required by the protocol.

glbrntt commented on Sources/GRPC/GRPCProtobufPayload.swift:
```diff
@@ -0,0 +1,32 @@
+
+public protocol GRPCProtobufPayload: GRPCPayload {}
+
+extension GRPCProtobufPayload where Self: SwiftProtobuf.Message {
```
The extension needs to be `public` as well so that it's visible in the module containing the generated code (i.e. outside of `GRPC`).

mustiikhalil commented:
@glbrntt I think this is ready for review now! Thanks a million for the guidance though, did learn now things!

glbrntt commented on Sources/protoc-gen-grpc-swift/Generator.swift:
```diff
@@ -117,5 +117,23 @@ class Generator {
+  }
+    
+  internal func printProtoBufExtensions() {
+    println("/// Usage: Confirms protoc objects to  GRPCProtoBufPayload")
```
```suggestion
    println("/// Provides conformance to `GRPCPayload` for the request and response messages")
```

glbrntt commented on Sources/protoc-gen-grpc-swift/Generator.swift:
```diff
@@ -117,5 +117,23 @@ class Generator {
       }
     }
+    println()
+    printProtoBufExtensions()
```
You need to call this for each service in the file as well.

glbrntt commented on Sources/protoc-gen-grpc-swift/Generator.swift:
```diff
@@ -117,5 +117,23 @@ class Generator {
+    }
+  }
+
+  private func printExtension(str: String, set: inout Set<String>) {
```
Perhaps something more readable, like:

```swift
private func printExtension(for messageType: String, typesSeen: inout Set<String>)
```

glbrntt commented on Sources/GRPC/GRPCProtobufPayload.swift:
```diff
@@ -0,0 +1,32 @@
+                  partial: false,
+                  options: BinaryDecodingOptions())
+    }
+    func serialize(into buffer: inout NIO.ByteBuffer) throws {
```
Can you add a blank line between these?

glbrntt commented on Sources/GRPC/GRPCPayload.swift:
```diff
@@ -0,0 +1,21 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
```
nit: can you reindent this file with 2 spaces please?

glbrntt commented on Sources/GRPC/GRPCProtobufPayload.swift:
```diff
@@ -0,0 +1,32 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
```
nit: can you reindent this file with 2 spaces please?

mustiikhalil commented on Sources/GRPC/GRPCServerCodec.swift:
```diff
@@ -75,8 +74,9 @@ extension GRPCServerCodec: ChannelOutboundHandler {
-        context.write(self.wrapOutboundOut(.message(messageData)), promise: promise)
+        var buffer = ByteBufferAllocator().buffer(capacity: 0)
+        try message.serialize(into: &buffer)
+        context.write(self.wrapOutboundOut(.message(buffer.readData(length: buffer.readableBytes)!)), promise: promise)
```
I would gladly do it but do you have a structure in mind or should I look into how we can do it?

glbrntt replied:
Sure, I'll try to explain the problem and what needs to be done to improve it.

### Problem

As you know, before this change we explicitly expected payloads to be `Message` from SwiftProtobuf. The only serialization and deserialization methods from `Message` are via `Data` from Foundation. Internally we use NIO and our currency type for storing bytes is NIOs `ByteBuffer`.

Because we always expect a `Data` from `Message`, the `write` method on `LengthPrefixedMessageWriter` accepts a `Data` and writes into a `ByteBuffer`.

With the protocol we always serialize into a `ByteBuffer` (for `Message` we _have_ to go via `Data` first). This means that we've gone from:

1. `Message` to `Data` (serialization)
2. `Data` to `ByteBuffer` (length prefixing)

to

1. `Message` to `Data` (serialization) 
2. `Data` to `ByteBuffer` (we need to for `GRPCPayload`)
3. `ByteBuffer` to `Data` (we need to for `LengthPrefixedMessageWriter.write`)
4. `Data` to `ByteBuffer` (length prefixing)

Clearly we don't want to copy bytes from `Data` to `ByteBuffer` and back from `ByteBuffer` to `Data` again (and we also don't want our API for `GRPCPayload` to be defined by `Data`).

Now I need to provide some background on the `LenthPrefixedMessageWriter`: gRPC payloads are sent with a 5-byte prefix (I might have explained this before but I'll repeat it for completeness). The first byte indicates whether the payload is compressed or not, the next four bytes indicate the length of the payload (the compressed length if it is compressed). 

If the payload isn't compressed then this is actually really straightforward: we write a `0`-byte we write the length of the payload as a `UInt32` and then we write the payload. For compression we have to write the a `1`-byte, leave a 4-byte gap for the length, compress the payload into the buffer and not how many bytes it was and then fill in the 4-byte gap.

### Possible Solution

So I think the best we can do now (I'll use protobuf as an example again) is actually to pre-allocate the extra 5-byte space at the beginning of the buffer before we pass it into `message.serialize(into:)`.

If we aren't using compression then all we need to do is fill in the first 5 bytes, easy! (And we have the `Message` to `Data` to `ByteBuffer`, as we had before, for Flatbuffers I expect we have something similar.)

When we use compression things are slightly different because we can't compress in place, so we have an extra buffer but in this case the same applies, we can avoid the extra hop via `Data`.

Hopefully this all makes sense so far?

### What needs to be done?

- `LengthPrefixedMessageWriter.write` accepts a `Data`, instead I think the writer should accept a `GRPCPayload` (as the writer knows whether compression is being used so knows how to fill in the first 5-bytes of the payload etc, it can also choose whether it _should_ preallocate 5 bytes, if we're using compression then we don't need to since we have to compress into a different buffer anyway).
- `Zlib.deflate` also accepts a `Data` if I recall correctly, this will need to be changed to a `ByteBuffer`

On the server side of things, this will require having to move some components around. At the moment the length prefixed writer lives in `HTTP1ToRawGRPCServerCodec`; this will have to move to `GPRCServerCodec`. This means the types that get sent between these handlers will also have to change (`_RawGRPCServerResponsePart.message` will have a `ByteBuffer` as its payload which will be a length prefixed message). I also expect a number of tests will have to be updated to support these changes.

mustiikhalil commented:
I've fixed all the suggested things, and also refactored the code according to what we planned yesterday, however there is a point in the `HTTP1ToRawGRPCServerCodec` where I just disregarded the first 5 bytes in the buffer, which I really hate doing, but I simply don't fully understand the library to come up with a solution. Furthermore, I ran both commands locally and there is no difference between both I am not sure why the CLI is having issues with it :/ 
```
protoc Sources/Examples/Echo/Model/echo.proto \                                                                                                              --proto_path=./Sources/Examples/Echo/Model/ \
--plugin=./.build/release/protoc-gen-grpc-swift \
--grpc-swift_opt=Visibility=Public \
--grpc-swift_out=.

diff -u ./echo.grpc.swift Sources/Examples/Echo/Model/echo.grpc.swift
```

glbrntt commented:
Great, I'll take another look now.

> Furthermore, I ran both commands locally and there is no difference between both I am not sure why the CLI is having issues with it :/
> (Yes, the generated file is built by the cli)
> 
> ```
> protoc Sources/Examples/Echo/Model/echo.proto \                                                                                                              --proto_path=./Sources/Examples/Echo/Model/ \
> --plugin=./.build/release/protoc-gen-grpc-swift \
> --grpc-swift_opt=Visibility=Public \
> --grpc-swift_out=.
> 
> diff -u ./echo.grpc.swift Sources/Examples/Echo/Model/echo.grpc.swift
> ```

Oh I think I see the problem, the CI looks like it isn't actually rebuilding the plugin (maybe the workspace gets cached between different commits for the same PR?!). If you update the rule in the `Makefile` for building the plugin (line 34) to be the following then it should work 🤞:

```make
${PROTOC_GEN_GRPC_SWIFT}: Sources/protoc-gen-grpc-swift/*.swift
```

mustiikhalil commented:
let's see hopefully it works.

glbrntt commented on Sources/GRPCInteroperabilityTestModels/Generated/test.grpc.swift:
```diff
@@ -361,3 +361,21 @@ extension Grpc_Testing_ReconnectServiceProvider {
   }
 }
 
+extension Grpc_Testing_SimpleRequest: GRPCProtobufPayload {}
```
Was this regenerated? (Doesn't have the `/// Provides conformance to ...`)

glbrntt commented on Sources/GRPC/LengthPrefixedMessageWriter.swift:
```diff
@@ -44,44 +44,58 @@ internal struct LengthPrefixedMessageWriter {
       buffer.moveWriterIndex(forwardBy: MemoryLayout<UInt32>.size)
 
+      // Build the message by serializing it into a Buffer first
+      // Since deflate can only take a ByteBuffer
```
`deflate` can only take a `ByteBuffer` because we decided it should be that way. I don't think we need a comment here to be honest though.

glbrntt commented on Sources/GRPC/GRPCServerCodec.swift:
```diff
@@ -75,8 +78,9 @@ extension GRPCServerCodec: ChannelOutboundHandler {
       do {
-        let messageData = try message.serializedData()
-        context.write(self.wrapOutboundOut(.message(messageData)), promise: promise)
+        var responseBuffer = context.channel.allocator.buffer(capacity: protobufMetadataSize)
```
I think you can just allocate an empty buffer here, the `writer` reserves the space we need.

glbrntt commented on Sources/GRPC/GRPCServerCodec.swift:
```diff
@@ -75,8 +78,9 @@ extension GRPCServerCodec: ChannelOutboundHandler {
-        let messageData = try message.serializedData()
-        context.write(self.wrapOutboundOut(.message(messageData)), promise: promise)
+        var responseBuffer = context.channel.allocator.buffer(capacity: protobufMetadataSize)
+        try messageWriter.write(message, into: &responseBuffer)
```
nit: we use explicit `self.` everywhere (it makes things easier to review since it's more obvious where something is coming from)

mustiikhalil commented on Sources/GRPC/LengthPrefixedMessageWriter.swift:
```diff
@@ -44,44 +44,58 @@ internal struct LengthPrefixedMessageWriter {
+      
+      // Leave a gap for the length, we'll set it in a moment.
+      let payloadSizeIndex = buffer.writerIndex
+      buffer.writeInteger(UInt32(0))
```
I've actually tried to use `moveWriterIndex` before but since we don't know the size of the message (Payload) anymore we can't really allocate the required space by using the minimum `reserveCapacity`

mustiikhalil commented on Makefile:
```diff
@@ -31,8 +31,7 @@ plugins: ${PROTOC_GEN_SWIFT} ${PROTOC_GEN_GRPC_SWIFT}
 	${SWIFT_BUILD_RELEASE} --product protoc-gen-swift
 
-${PROTOC_GEN_GRPC_SWIFT}:
-	${SWIFT_BUILD_RELEASE} --product protoc-gen-grpc-swift
```
I've tried everything in here but with no luck the CLI keeps on erring out

glbrntt replied:
Did you try this?

```make
${PROTOC_GEN_GRPC_SWIFT}: Sources/protoc-gen-grpc-swift/*.swift
	${SWIFT_BUILD_RELEASE} --product protoc-gen-grpc-swift
```

glbrntt commented on Sources/GRPC/Compression/Zlib.swift:
```diff
@@ -49,24 +49,27 @@ enum Zlib {
+          inputBuffer: CGRPCZlib_castVoidToBytefPointer(inputPointer.baseAddress!),
+          inputBufferSize: inputPointer.count
+        )
+      }
```
This isn't quite right: the input buffer pointer is now escaping the `input.readWithUnsafeMutableReadableBytes` closure (since it's assigned to the input buffer of `self.stream` which outlives the closure.

You should follow the pattern used for `inflate`; it should be something like:

```swift
fun deflate(...) -> Int {
  let upperBound =  ...
  
  return try input.readWithUnsafeMutableReadableBytes { inputPointer -> (Int, Int) in 
    self.stream.nextInputBuffer = ...
    self.stream.availableInputBytes = ... 

    defer {
      self.stream.nextInputBuffer = nil
      self.stream.availableInputBytes = 0
    }

    return try output.writeWithUnsafeMutableBytes(...) {
      ...
    }
}
```

mustiikhalil replied:
I'm still trying to fix this

glbrntt commented on Sources/GRPC/LengthPrefixedMessageWriter.swift:
```diff
@@ -44,25 +44,27 @@ internal struct LengthPrefixedMessageWriter {
-    buffer.reserveCapacity(LengthPrefixedMessageWriter.metadataLength + message.count)
-
+  func write(_ payload: GRPCPayload, into buffer: inout ByteBuffer, disableCompression: Bool = false) throws {
+    buffer.reserveCapacity(MemoryLayout.size(ofValue: payload) + LengthPrefixedMessageWriter.metadataLength)
```
`MemoryLayout.size(ofValue: payload)` doesn't actually help us here since it isn't the serialised size of the payload. Just the metadata length is absolutely fine here.

mustiikhalil replied:
@glbrntt I actually used this because there was a test case crash where ReserveCapacity won’t really reserve enough bytes and the bytebuffer would have a precondition crash while trying to move the writer. Hence, the earlier code where I would write the bytes and then go back and replace them

glbrntt commented on Sources/protoc-gen-grpc-swift/Generator.swift:
```diff
@@ -117,5 +117,27 @@ class Generator {
+    var writtenValues = Set<String>()
+    for service in file.services {
+      self.service = service
+      println("/// Provides conformance to `GRPCPayload` for the request and response messages")
```
Can you move this up before the `services` loop?

glbrntt commented on Sources/GRPC/HTTP1ToRawGRPCServerCodec.swift:
```diff
@@ -260,27 +256,21 @@ extension HTTP1ToRawGRPCServerCodec: ChannelOutboundHandler {
-        try! messageWriter.write(messageBytes, into: &self.responseTextBuffer)
-
+        
+        self.responseTextBuffer.writeBytes(messageBytes.readableBytesView)
```
```suggestion
        self.responseTextBuffer.writeBuffer(messageBytes)
```

glbrntt commented on Sources/GRPC/GRPCPayload.swift:
```diff
@@ -0,0 +1,33 @@
+  /// - Throws: Should throw an error if the data wasn't serialized properly
+  init(serializedByteBuffer: inout NIO.ByteBuffer) throws
+
+  /// serializes data into a `ByteBuffer`
```
```suggestion
  /// Serializes the payload into a `ByteBuffer`.
```

glbrntt commented on Sources/GRPC/GRPCPayload.swift:
```diff
@@ -0,0 +1,33 @@
+  
+  /// Initializes a new payload object from a given `NIO.ByteBuffer`
+  ///
+  /// - Parameter serializedByteBuffer:
```
```suggestion
  /// - Parameter serializedByteBuffer: A buffer containing the serialised bytes of this payload.
```

glbrntt commented on Sources/GRPC/GRPCPayload.swift:
```diff
@@ -0,0 +1,33 @@
+  ///
+  /// - Parameters:
+  ///   - buffer: The buffer to write the payload into.
+  /// - Note: Implementer should NOT clear `ByteBuffer`
```
```suggestion
  /// - Note: Implementers must *NOT* clear or read bytes from `buffer`.
```

mustiikhalil commented on Sources/GRPC/Compression/Zlib.swift:
```diff
@@ -69,12 +67,15 @@ enum Zlib {
         }
+        
+        let bytesRead = inputPointer.count - self.stream.availableInputBytes
+        return (bytesRead, writtenBytes)
```
@glbrntt I think this should fix the issue with the pointer being accessed while we already moved out of the function

mustiikhalil commented on Sources/GRPC/LengthPrefixedMessageWriter.swift:
```diff
@@ -49,9 +49,8 @@ internal struct LengthPrefixedMessageWriter {
-    buffer.reserveCapacity(MemoryLayout.size(ofValue: payload) + LengthPrefixedMessageWriter.metadataLength)
-    
+  func write(_ payload: GRPCPayload, into buffer: inout ByteBuffer, disableCompression: Bool = false) throws {r
+    buffer.reserveCapacity(buffer.writerIndex + LengthPrefixedMessageWriter.metadataLength)
```
moved the code to use the `buffer.writerIndex` instead of the `MemoryLayout.size(of: payload)` since I think this way it will be more robust knowing that it will always have those 5 extra bytes on the buffer.

glbrntt replied:
Great catch, thanks!

mustiikhalil commented:
Wonderful!
